### PR TITLE
Enhance OpenID Connect Plugin

### PR DESCRIPTION
**Pull request by dr4hcu5-jan**
This PR enhances the maintainability of the OpenID Connect plugin which is used to authenticate requests to the microservices. Furthermore, the Docker image build has been made more-caching friendly to allow for a better deployment experience

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,23 +1,28 @@
...
+> This API Gateway is based on the Open Source Edition of the [Kong API Gateway](https://konghq.com/products/kong-gateway).
+> Therefore, features available in the free (closed-source) and enterprise (also
+> closed-source) edition are not available with this image.
 
-- [x] OpenID Connect Authentication/Authorization with JWT Validation
+The API Gateway is the central part of the WISdoM Plaform and manages the access
+to the microservices implemented for the platform.
+Furhtermore, it is extended by a plugin using the 
+[`go-pdk`](https://github.com/Kong/go-pdk) to validate JWTs which are used in a
+standard deplyoment to authenticate and secure requests.
```
Comment (cptpiepmatz) [inline README.md]:
```suggestion
standard deployment to authenticate and secure requests.
```

File: README.md [hunk 2]
```diff
@@ -1,23 +1,28 @@
...
-API Gateway.
-
-## How to use
-Since the API gateway in included in every deployment of the WISdoM platform,
-there are no extra steps you need to take.
\ No newline at end of file
+The API Gateway is autommatically configured using the 
+[gateway-service-watcher](https://github.com/wisdom-oss/gateway-service-watcher)
+which acts as a watchdog to check the deployed containers on the host for their
+association to the WISdoM plaform.
```
Comment (cptpiepmatz) [inline README.md]:
```suggestion
association to the WISdoM platform.
```

File: plugins/oidc/access.go [hunk 3]
```diff
@@ -10,118 +10,207 @@ import (
...
+	res, err := http.Get(discoveryEndpoint)
+	if err != nil {
+		return "", "", errors.Join(errDiscoveryRequestFailure, err)
+	}
+
+	// now read the openid connect configuration
+	var openIdConnectConfiguration map[string]interface{}
+	err = json.NewDecoder(res.Body).Decode(&openIdConnectConfiguration)
+	if err != nil {
+		return "", "", errors.Join(err, errDiscoveryUnmarshalFailure)
```
Comment (cptpiepmatz) [inline plugins/oidc/access.go:86]: I think I would return `nil` instead of empty strings in the error case, but that is not really an issue.

File: plugins/oidc/access.go [hunk 4]
```diff
@@ -10,118 +10,207 @@ import (
...
+	}
+
+	return openIdConnectConfiguration["userinfo_endpoint"].(string), openIdConnectConfiguration["jwks_uri"].(string), nil
```
Comment (cptpiepmatz) [inline plugins/oidc/access.go]: The assertion here would crash the application, is that expected behavior?
Maybe do a type check and return an error if that the json has an incorrect format instead.

File: plugins/oidc/access.go [hunk 5]
```diff
@@ -10,118 +10,207 @@ import (
...
+//	// token = "abcxyz123"
+//
+// Parameters:
+//   - authorizationHeaderValue: The value of the authorization header.
+//
+// Returns:
+//
+//	The extracted bearer token, or an empty string if no valid bearer token is found.
+func extractBearerToken(authorizationHeaderValue string) string {
+	regex := regexp.MustCompile(headerRegex)
```
Comment (cptpiepmatz) [inline plugins/oidc/access.go]: Why did you move the regex compile step into the function, will that be rerun every time?

File: plugins/oidc/access.go [hunk 6]
```diff
@@ -10,118 +10,207 @@ import (
...
+	matches := regex.FindStringSubmatch(authorizationHeaderValue)
+	if len(matches) == 3 {
+		return matches[2]
+	}
```
Comment (cptpiepmatz) [inline plugins/oidc/access.go:132]: This works but a [named group](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Named_capturing_group) could be clearer to use.

File: plugins/oidc/access.go [hunk 7]
```diff
@@ -10,118 +10,207 @@ import (
...
+	return ""
```
Comment (cptpiepmatz) [inline plugins/oidc/access.go:133]: Why return an empty string instead of `nil`. I think a `nil` would be clearer. (Or some optional type.)

File: plugins/oidc/access.go [hunk 8]
```diff
@@ -10,118 +10,207 @@ import (
...
+}
+
+// Access is executed every time the kong gateway receives a request. Since ther
```
Comment (cptpiepmatz) [inline plugins/oidc/access.go:136]:
```suggestion
// Access is executed every time the kong gateway receives a request. Since the
```

File: plugins/oidc/access.go [hunk 9]
```diff
@@ -10,118 +10,207 @@ import (
...
-	authorizationHeader, _ := request.GetHeader("Authorization")
-	// now trim away any blank values at the start or end of the string
-	authorizationHeader = strings.TrimSpace(authorizationHeader)
-	// now get the authorization method and token from the header
-	matches := headerRegex.FindStringSubmatch(authorizationHeader)
-	if len(matches) != 3 {
-		// since the matches array needs to contain three entries, the value is
-		// invalid and therefore the user is not authorized
+
+	userinfoEndpoint, jwksEndpoint, err := discoverEndpoints(c.DiscoveryUri)
```
Comment (cptpiepmatz) [inline plugins/oidc/access.go:142]: The endpoint discovery always runs a GET request which takes some time. Maybe this is a bit too rigorous.

File: plugins/oidc/access.go [hunk 10]
```diff
@@ -10,118 +10,207 @@ import (
...
+				ErrorCode:        "gateway.OPENID_CONNECT_DISCOVERY_REQUEST_FAILURE",
+				ErrorTitle:       "Discovery Request Failure",
+				ErrorDescription: "The discovery request needed for access token validation failed: " + err.Error(),
+			}
+		case errors.Is(err, errDiscoveryUnmarshalFailure):
+			response = GatewayError{
+				ErrorCode:        "gateway.OPENID_CONNECT_DISCOVERY_RESPONSE_PARSE_ERROR",
+				ErrorTitle:       "Discovery Response Parsing Error",
+				ErrorDescription: "The discovery response sent by the configured endpoint could not be parsed: " + err.Error(),
+			}
```
Comment (cptpiepmatz) [inline plugins/oidc/access.go:167]: This is nice :+1:

File: plugins/oidc/access.go [hunk 11]
```diff
@@ -10,118 +10,207 @@ import (
...
+	matches := regex.FindStringSubmatch(authorizationHeaderValue)
+	if len(matches) == 3 {
+		return matches[2]
+	}
+	return ""
+}
+
+// Access is executed every time the kong gateway receives a request. Since ther
 // plugin may be restarted at any moment
 func (c *Configuration) Access(kong *pdk.PDK) {
```
Comment (cptpiepmatz) [inline plugins/oidc/access.go:138]: This entire function is very long. You have multiple phases/steps. Maybe reduce this function to calls into subfunctions that each handle the steps. 

Maybe you can also shorten the error handling by constructing the error type in the subfunctions, by this you can something more like this:
```go
validated, err := validateSomething()
if err != nil {
  return nil, err
}
```

**Review by cptpiepmatz — CHANGES_REQUESTED** [review]

File: plugins/oidc/access.go (hunk 5 above)
Comment (dr4hcu5-jan) [reply plugins/oidc/access.go]: The compile step has been moved down here because sometimes the regex compilation failed. However, the regex compile step can be made to be only executed once, using the following code:
```go
// at the top of the file
const bearerTokenExtractionRegExString = `^(\w+) (\S+)$`
var bearerTokenExtractionRegEx *regexp.Regexp

func extractBearerToken(authorizationHeaderValue string) string {
    if bearerTokenExtractionRegEx == nil {
      bearerTokenExtractionRegEx = regexp.MustCompile(bearerTokenExtractionRegExString)
    }
    // rest of the function...
}
```
This way the regex will only compiled once, but only if the function is called

File: plugins/oidc/access.go (hunk 6 above)
Comment (dr4hcu5-jan) [reply plugins/oidc/access.go:132]: Using a named group will only make the regex more illegible since there is no function in the standard package for regular expressions that handles returning named groups directly. Therefore, in my opinion using a named group will not increase the clarity of this part
> https://pkg.go.dev/regexp#pkg-index

File: plugins/oidc/access.go (hunk 7 above)
Comment (dr4hcu5-jan) [reply plugins/oidc/access.go:133]: Returning `nil` would require using `*string` as return type and requires a more thorough check of the return type in the code later on, since a `*string` cannot be used directly in a `if` statement to check the length and would require the following syntax making the function using the return value longer:
```go
if accessToken == nil {
  // raise error here
}

if strings.Trim(*accessToken) == "" {
  // raise another error here
}
```

File: plugins/oidc/access.go (hunk 9 above)
Comment (dr4hcu5-jan) [reply plugins/oidc/access.go:142]: True, however this is the only way a configuration change during the plugin runtime can be caught since the plugin does not need to be restarted if the plugin configuration has been changed and caching these values may result in an validation error further down in the code

File: plugins/oidc/access.go (hunk 4 above)
Comment (dr4hcu5-jan) [reply plugins/oidc/access.go]: Yes, this type assertion would crash the application. A type check will be added

File: plugins/oidc/access.go (hunk 3 above)
Comment (dr4hcu5-jan) [reply plugins/oidc/access.go:86]: This is the same as the following comment: https://github.com/wisdom-oss/api-gateway/pull/4#discussion_r1470902455

File: plugins/oidc/access.go (hunk 5 above)
Comment (cptpiepmatz) [reply plugins/oidc/access.go]: This is a good alternative. But why would a hard-coded regex fail to compile sometimes?

File: plugins/oidc/README.md [hunk 12]
```diff
@@ -20,5 +20,7 @@ These responses can optionally be cached using a redis database to further
 minimize the number of network requests executed for each intercepted request.
 
 ## Configuration
-The configuration is done automatically during the startup of the api gateway
-using the environment variables passed to the api gateway
\ No newline at end of file
+The configuration is done by the 
+[watchdog](https://github.com/wisdom-oss/watchdog) on a service-by-service
+basis.
+Please refer to the documentation of the watchdog for further information
```
Comment (cptpiepmatz) [inline plugins/oidc/README.md]:
```suggestion
Please refer to the documentation of the watchdog for further information.
```

File: plugins/oidc/access.go (hunk 5 above)
Comment (dr4hcu5-jan) [reply plugins/oidc/access.go]: Local testing showed that not the regex failed to compile, but trying to use the regex resulted in a panic due to a nil pointer dereference. This happened if there was a request in the backlog of the gateway and the plugin is not started fully which happens if the configuration of the plugin has been changed. This is an edge case and was only reproduced by loading the gateway with about 25000 requests in its backlog and changing the gateway every 2.5 seconds.

Comment (cptpiepmatz) [reply plugins/oidc/access.go]: Huh, then we could just move the regex compile up again or not?

Comment (dr4hcu5-jan) [reply plugins/oidc/access.go]: No, because using the compile step in this way ensures that the regex is not `nil` before usage and if the regex is `nil` it compiles the regex

Comment (dr4hcu5-jan) [reply plugins/oidc/access.go]: Furthermore, the regex is now compiled only once and not at every access which reduces the time for handling the request

**Review by cptpiepmatz — APPROVED** [review]
